fdbclient/FileBackupAgent.actor.cpp

```diff
@@ -458,8 +475,417 @@ Value makePadding(int size) {
+
+		// write options struct
+		Value serialized =
+		    ObjectWriter::toValue(self->options, IncludeVersion(ProtocolVersion::withEncryptedSnapshotBackupFile()));
+		appendStringRefWithLenToBuffer(self, &serialized);
+
+		// leave space for encryption header
+		self->encryptHeader = (BlobCipherEncryptHeader*)self->wPtr;
+		self->wPtr += BlobCipherEncryptHeader::headerSize;
+		self->dataPayloadStart = self->wPtr;
+
+		// If this is NOT the first block then write duplicate stuff needed from last block
+		if (self->blockEnd > self->blockSize) {
+			appendStringRefWithLenToBuffer(self, &lastKey);
+			appendStringRefWithLenToBuffer(self, &self->lastKey);
+			if (writeValue) {
+				appendStringRefWithLenToBuffer(self, &self->lastValue);
+			}
+		}
+
+		// There must now be room in the current block for bytesNeeded or the block size is too small
+		if (expectedFileSize(self) + bytesNeeded > self->blockEnd) {
+			throw backup_bad_block_size();
+		}
+
+		return Void();
+	}
+
+	Future<Void> padEnd(bool final) {
+		if (expectedFileSize(this) > 0) {
+			return newBlock(this, 0, StringRef(), true, final);
+		}
+		return Void();
+	}
+
+	// Ends the current block if necessary based on bytesNeeded.
+	ACTOR static Future<Void> newBlockIfNeeded(EncryptedRangeFileWriter* self, int bytesNeeded) {
+		if (expectedFileSize(self) + bytesNeeded > self->blockEnd) {
+			wait(newBlock(self, bytesNeeded, self->lastKey, true));
+		}
+		return Void();
+	}
+
+	ACTOR static Future<bool> finishCurTenantBlockStartNewIfNeeded(EncryptedRangeFileWriter* self,
+	                                                               Key k,
+	                                                               Value v,
+	                                                               bool writeValue) {
+		// Don't want to start a new block if the current key or previous key is empty
+		if (self->lastKey.size() == 0 || k.size() == 0) {
+			return false;
+		}
+		state std::pair<int64_t, TenantName> curKeyTenantInfo = wait(getTenantIdName(k, self->tenantCache));
+		state std::pair<int64_t, TenantName> prevKeyTenantInfo =
+		    wait(getTenantIdName(self->lastKey, self->tenantCache));
+		// crossing tenant boundaries so finish the current block using only the tenant prefix of the new key
+		if (curKeyTenantInfo.first != prevKeyTenantInfo.first) {
+			CODE_PROBE(true, "crossed tenant boundaries");
+			state KeyRef endKey = k;
+			// If we are crossing a boundary with a key that has a tenant prefix then truncate it
+			if (curKeyTenantInfo.first != SYSTEM_KEYSPACE_ENCRYPT_DOMAIN_ID &&
+			    curKeyTenantInfo.first != FDB_DEFAULT_ENCRYPT_DOMAIN_ID) {
+				endKey = StringRef(k.begin(), 8);
+			}
+			state ValueRef newValue = StringRef();
+			self->lastKey = k;
+			self->lastValue = v;
+			appendStringRefWithLenToBuffer(self, &endKey);
+			appendStringRefWithLenToBuffer(self, &newValue);
+			wait(newBlock(self, 0, endKey, writeValue));
+			wait(updateEncryptionKeys(self, self->lastKey, self->tenantCache));
```

Comment:
Few thoughts:
1. Better to rename method to `updateEncryptionKeysCtx` as it updates: `textCipherKey`, `headerCipherKey` and `iv`.
2. Shouldn't this be done as part of initializing the `nextBlock` instead of `finalizing the current block`. Subtle difference, but the current code may end up fetching the keys and generating IV even for the very last block flushed as part of this snapshot..

Reply:
Actually for the very last block we will only refresh the encryption key if the last key is in a different tenant (expected behaviour), see the `writeKey_impl` method. We only ever call this method in two cases:

1.) There are no encryption keys currently present (first time we get a call to `writeKey_impl` or `writeKV_impl`)
2.) We are switching tenant boundaries (in which case we will always need to switch keys)

Reply:
My point is why to refresh encryption key in the last block, shouldn't it be done `only` while appending first KV to the newBlock? It could also simplify `key-refresh` logic as we don't need to refresh encryption keys in the middle of a block.

Reply:
If you look one line above you can see we are creating a new block. So this is already doing what you want (only refreshing the keys) after a new block. This only ever gets called when we switch from one tenant to another

Reply:
I think the confusion here is maybe that this part of the code is only ever entered when tenant boundaries are crossed

Reply:
I see, I would rathe abstract the code into routine maybe named as `handleTenantBondary()`, it would make code more readable. Thoughts?